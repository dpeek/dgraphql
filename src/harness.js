import fs from 'fs'
import path from 'path'

import { graphql } from 'graphql'

import { buildSchema } from './schema'

import type { ClientConfig } from './client'

export async function init (config: ClientConfig) {
  const schemaPath = path.resolve(__dirname, '__tests__/schema.graphql')
  const sourcePath = path.resolve(__dirname, '__tests__/source.graphql')

  const source = fs.readFileSync(sourcePath).toString()

  const time = String(new Date().getTime() - 1495660000000)
  let commonVariables = { time }

  // can't use the same schema as some tests are in relay mode, the query is not
  const dataSchema = buildSchema(fs.readFileSync(schemaPath).toString(), {})
  const data = await graphql({
    schema: dataSchema,
    source,
    variableValues: commonVariables,
    contextValue: { language: 'en' }
  })
  const walk = node => {
    if (node.id && node.name) {
      commonVariables[node.name.split(' ')[0].toLowerCase()] = node.id
    }
    Object.values(node).forEach(value => {
      if (typeof value === 'object') walk(value)
    })
  }
  walk(data)

  const schema = buildSchema(fs.readFileSync(schemaPath).toString(), config)
  return (source, variables, language) => {
    return graphql({
      schema,
      source,
      variableValues: { ...commonVariables, ...(variables || {}) },
      contextValue: { language: language || 'en' }
    })
  }
}
