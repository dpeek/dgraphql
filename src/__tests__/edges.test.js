import { init } from './harness'

var sequence

beforeAll(async () => {
  const test = await init()
  sequence = test.sequence
})

test('sets node edge to existing node', () => {
  const query1 = `mutation Test {
    tim: createPerson(input: {
      name: "Tim"
    }) {
      person {
        id
        name
      }
    }
    alice: createPerson(input: {
      name: "Alice"
    }) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `mutation Test ($tim: ID!, $alice: ID!) {
    setPersonPartner(input: {
      id: $tim,
      partner: {
        id: $alice
      }
    }) {
      person {
        name
        partner {
          name
          partner {
            name
          }
        }
      }
    }
  }`
  return sequence([query1, query2])
})

test('sets one way edge to existing node', () => {
  const query1 = `mutation Test {
    tim: createPerson(input: {
      name: "Tim"
    }) {
      person {
        id
        name
      }
    }
    alice: createPerson(input: {
      name: "Alice"
    }) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `mutation Test ($tim: ID!, $alice: ID!) {
    setPersonEmergencyContact(input: {
      id: $tim,
      emergencyContact: {
        id: $alice
      }
    }) {
      person {
        name
        emergencyContact {
          name
        }
      }
    }
  }`
  return sequence([query1, query2])
})

test('sets node edge to new node', () => {
  const query1 = `mutation Test {
    tim: createPerson(input: {
      name: "Tim"
    }) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `mutation Test ($tim: ID!) {
    setPersonPartner(input: {
      id: $tim,
      partner: {
        name: "Wilma"
      }
    }) {
      person {
        name
        partner {
          name
          partner {
            name
          }
        }
      }
    }
  }`
  return sequence([query1, query2])
})

test('setting node edge to existing value has no effect', () => {
  const query1 = `mutation Test {
    tim: createPerson(input: {
      name: "Tim",
      partner: {
        name: "Alice"
      }
    }) {
      person {
        id
        name
        partner {
          id
          name
        }
      }
    }
  }`
  const query2 = `mutation Test ($tim: ID!, $alice: ID!) {
    setPersonPartner(input: {
      id: $tim,
      partner: {
        id: $alice
      }
    }) {
      person {
        name
        partner {
          name
          partner {
            name
          }
        }
      }
    }
  }`
  return sequence([query1, query2])
})

test('setting edge with reverse removes reverse edge from existing node', () => {
  const query1 = `mutation Test {
    tim: createPerson(input: {
      name: "Tim",
      partner: {
        name: "Sally"
      }
    }) {
      person {
        id
        name
        partner {
          id
          name
        }
      }
    }
    alice: createPerson(input: {
      name: "Alice"
    }) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `mutation Test ($tim: ID!, $alice: ID!) {
    setPersonPartner(input: {
      id: $tim,
      partner: {
        id: $alice
      }
    }) {
      person {
        name
        partner {
          name
        }
      }
    }
  }`
  const query3 = `query Test ($sally: ID!) {
    person(id: $sally) {
      name
      partner {
        id
      }
    }
  }`
  return sequence([query1, query2, query3])
})

test('setting edge with no reverse does nothing to existing node', () => {
  const query1 = `mutation Test {
    tim: createPerson(input: {
      name: "Tim",
      emergencyContact: {
        name: "Sally",
        emergencyContact: {
          name: "Gerald"
        }
      }
    }) {
      person {
        id
        name
        emergencyContact {
          id
          name
        }
      }
    }
    alice: createPerson(input: {
      name: "Alice"
    }) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `mutation Test ($tim: ID!, $alice: ID!) {
    setPersonEmergencyContact(input: {
      id: $tim,
      emergencyContact: {
        id: $alice
      }
    }) {
      person {
        name
        emergencyContact {
          name
        }
      }
    }
  }`
  const query3 = `query Test ($sally: ID!) {
    person(id: $sally) {
      name
      emergencyContact {
        name
      }
    }
  }`
  return sequence([query1, query2, query3])
})

test('setting node edge removes existing reverse edge from new value', () => {
  const query1 = `mutation Test {
    tim: createPerson(input: {
      name: "Tim"
    }) {
      person {
        id
        name
      }
    }
    alice: createPerson(input: {
      name: "Alice",
      partner: {
        name: "Sue"
      }
    }) {
      person {
        id
        name
        partner {
          id
          name
        }
      }
    }
  }`
  const query2 = `mutation Test ($tim: ID!, $alice: ID!) {
    setPersonPartner(input: {
      id: $tim,
      partner: {
        id: $alice
      }
    }) {
      person {
        name
        partner {
          name
        }
      }
    }
  }`
  const query3 = `query Test ($sue: ID!) {
    person(id: $sue) {
      name
      partner {
        name
      }
    }
  }`
  return sequence([query1, query2, query3])
})

test('unsets node edge', () => {
  const query1 = `mutation Test {
    tim: createPerson(input: {
      name: "Tim",
      partner: {
        name: "Joan"
      }
    }) {
      person {
        id
        name
        partner {
          id
          name
        }
      }
    }
  }`
  const query2 = `mutation Test ($tim: ID!) {
    unsetPersonPartner(input: {
      id: $tim
    }) {
      person {
        name
        partner {
          name
        }
      }
    }
  }`
  return sequence([query1, query2])
})

test('unsetting node edge removes reverse edge on value', () => {
  const query1 = `mutation Test {
    tim: createPerson(input: {
      name: "Tim",
      partner: {
        name: "Joan"
      }
    }) {
      person {
        id
        name
        partner {
          id
          name
        }
      }
    }
  }`
  const query2 = `mutation Test ($tim: ID!) {
    unsetPersonPartner(input: { id: $tim }) {
      person {
        name
        partner {
          name
        }
      }
    }
  }`
  const query3 = `query Test ($joan: ID!) {
    person(id: $joan) {
      name
      partner {
        name
      }
    }
  }`
  return sequence([query1, query2, query3])
})

test('adds edges to existing nodes', () => {
  const query1 = `mutation Test {
    tim: createPerson(input: {
      name: "Tim"
    }) {
      person {
        id
        name
      }
    }
    alice: createPerson(input: {
      name: "Alice"
    }) {
      person {
        id
        name
      }
    }
    paula: createPerson(input: {
      name: "Paula"
    }) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `mutation Test ($tim: ID!, $alice: ID!, $paula: ID!) {
    addPersonChildren(input: {
      id: $tim,
      children: [{
        id: $alice
      },{
        id: $paula
      }]
    }) {
      person {
        name
        children(order: name_asc) {
          name
          parents {
            name
          }
        }
      }
    }
  }`
  return sequence([query1, query2])
})

test('adds edges to new nodes', () => {
  const query1 = `mutation Test {
    tim: createPerson(input: {
      name: "Tim"
    }) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `mutation Test ($tim: ID!) {
    addPersonChildren(input: {
      id: $tim,
      children: [{
        name: "John"
      },{
        name: "Jack"
      }]
    }) {
      person {
        name
        children(order: name_asc) {
          name
          parents {
            name
          }
        }
      }
    }
  }`
  return sequence([query1, query2])
})

test('adds edges to new and existing nodes', () => {
  const query1 = `mutation Test {
    tim: createPerson(input: {
      name: "Tim"
    }) {
      person {
        id
        name
      }
    }
    tom: createPerson(input: {
      name: "Tom"
    }) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `mutation Test ($tim: ID!, $tom: ID!) {
    addPersonChildren(input: {
      id: $tim,
      children: [{
        id: $tom
      },{
        name: "Eve"
      }]
    }) {
      person {
        name
        children(order: name_asc) {
          name
          parents {
            name
          }
        }
      }
    }
  }`
  return sequence([query1, query2])
})

test('adds edges of different type', () => {
  const query1 = `mutation Test {
    tim: createPerson(input: {
      name: "Tim"
    }) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `mutation Test ($tim: ID!) {
    addPersonEmails(input: {
      id: $tim,
      emails: [{
        type: HOME,
        address: "mail@tim.com"
      }]
    }) {
      person {
        emails {
          type
          address
        }
      }
    }
  }`
  return sequence([query1, query2])
})

test('adding edges does not remove existing edges', async () => {
  const query1 = `mutation Test {
    tim: createPerson(input: {
      name: "Tim",
      children: [{
        name: "Olivia"
      }]
    }) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `mutation Test($tim: ID!) {
    addPersonChildren(input: {
      id: $tim,
      children: [{
        name: "John"
      }]
    }) {
      person {
        name
        children(order: name_asc) {
          name
          parents {
            name
          }
        }
      }
    }
  }`
  return sequence([query1, query2])
})

test('removes edges removing reverse edges', async () => {
  const query1 = `mutation Test {
    tim: createPerson(input: {
      name: "Tim",
      children: [{
        name: "Jerry"
      }, {
        name: "Olivia"
      }, {
        name: "Paul"
      }]
    }) {
      person {
        id
        name
        children(order: name_asc) {
          id
          name
        }
      }
    }
  }`
  const query2 = `mutation Test($tim: ID!, $jerry: ID!, $paul: ID!) {
    removePersonChildren(input: {
      id: $tim,
      children: [{
        id: $jerry
      }, {
        id: $paul
      }]
    }) {
      person {
        name
        children(order: name_asc) {
          name
        }
      }
    }
  }`
  const query3 = `query Test($jerry: ID!) {
    person(id: $jerry) {
      name
      parents {
        name
      }
    }
  }`
  return sequence([query1, query2, query3])
})

test('setting edge to non-existent node returns error', async () => {
  const query1 = `mutation Test {
    tom: createPerson(input: {name: "Tom"}) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `mutation Test ($tom: ID!) {
    setPersonPartner(input: {id: $tom, partner: {id: "0x00"}}) {
      person {
        name
        partner {
          name
        }
      }
    }
  }`
  return sequence([query1, query2])
})

test('setting edge to existing node of incorrect type returns error', async () => {
  const query1 = `mutation Test {
    tom: createPerson(input: {name: "Tom"}) {
      person {
        id
        name
      }
    }
    tim: createPerson(input: {name: "Tim"}) {
      person {
        id
        name
      }
    }
  }`
  const query2 = `mutation Test($tim: ID!, $tom: ID!) {
    addPersonEmails(input: {id: $tom, emails: [{id: $tim}]}) {
      person {
        name
        partner {
          name
        }
      }
    }
  }`
  return sequence([query1, query2])
})
