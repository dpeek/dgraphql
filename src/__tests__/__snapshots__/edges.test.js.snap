// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adding edges does not remove existing edges 1`] = `
Object {
  "data": Object {
    "tim": Object {
      "person": Object {
        "name": "Tim",
      },
    },
  },
}
`;

exports[`adding edges does not remove existing edges 2`] = `
Object {
  "data": Object {
    "addPersonChildren": Object {
      "person": Object {
        "children": Object {
          "0": Object {
            "name": "John",
            "parents": Object {
              "0": Object {
                "name": "Tim",
              },
            },
          },
          "1": Object {
            "name": "Olivia",
            "parents": Object {
              "0": Object {
                "name": "Tim",
              },
            },
          },
        },
        "name": "Tim",
      },
    },
  },
}
`;

exports[`adds edges of different type 1`] = `
Object {
  "data": Object {
    "tim": Object {
      "person": Object {
        "name": "Tim",
      },
    },
  },
}
`;

exports[`adds edges of different type 2`] = `
Object {
  "data": Object {
    "addPersonEmails": Object {
      "person": Object {
        "emails": Object {
          "0": Object {
            "address": "mail@tim.com",
            "type": "HOME",
          },
        },
      },
    },
  },
}
`;

exports[`adds edges to existing nodes 1`] = `
Object {
  "data": Object {
    "alice": Object {
      "person": Object {
        "name": "Alice",
      },
    },
    "paula": Object {
      "person": Object {
        "name": "Paula",
      },
    },
    "tim": Object {
      "person": Object {
        "name": "Tim",
      },
    },
  },
}
`;

exports[`adds edges to existing nodes 2`] = `
Object {
  "data": Object {
    "addPersonChildren": Object {
      "person": Object {
        "children": Object {
          "0": Object {
            "name": "Alice",
            "parents": Object {
              "0": Object {
                "name": "Tim",
              },
            },
          },
          "1": Object {
            "name": "Paula",
            "parents": Object {
              "0": Object {
                "name": "Tim",
              },
            },
          },
        },
        "name": "Tim",
      },
    },
  },
}
`;

exports[`adds edges to new and existing nodes 1`] = `
Object {
  "data": Object {
    "tim": Object {
      "person": Object {
        "name": "Tim",
      },
    },
    "tom": Object {
      "person": Object {
        "name": "Tom",
      },
    },
  },
}
`;

exports[`adds edges to new and existing nodes 2`] = `
Object {
  "data": Object {
    "addPersonChildren": Object {
      "person": Object {
        "children": Object {
          "0": Object {
            "name": "Eve",
            "parents": Object {
              "0": Object {
                "name": "Tim",
              },
            },
          },
          "1": Object {
            "name": "Tom",
            "parents": Object {
              "0": Object {
                "name": "Tim",
              },
            },
          },
        },
        "name": "Tim",
      },
    },
  },
}
`;

exports[`adds edges to new nodes 1`] = `
Object {
  "data": Object {
    "tim": Object {
      "person": Object {
        "name": "Tim",
      },
    },
  },
}
`;

exports[`adds edges to new nodes 2`] = `
Object {
  "data": Object {
    "addPersonChildren": Object {
      "person": Object {
        "children": Object {
          "0": Object {
            "name": "Jack",
            "parents": Object {
              "0": Object {
                "name": "Tim",
              },
            },
          },
          "1": Object {
            "name": "John",
            "parents": Object {
              "0": Object {
                "name": "Tim",
              },
            },
          },
        },
        "name": "Tim",
      },
    },
  },
}
`;

exports[`removes edges removing reverse edges 1`] = `
Object {
  "data": Object {
    "tim": Object {
      "person": Object {
        "children": Object {
          "0": Object {
            "name": "Jerry",
          },
          "1": Object {
            "name": "Olivia",
          },
          "2": Object {
            "name": "Paul",
          },
        },
        "name": "Tim",
      },
    },
  },
}
`;

exports[`removes edges removing reverse edges 2`] = `
Object {
  "data": Object {
    "removePersonChildren": Object {
      "person": Object {
        "children": Object {
          "0": Object {
            "name": "Olivia",
          },
        },
        "name": "Tim",
      },
    },
  },
}
`;

exports[`removes edges removing reverse edges 3`] = `
Object {
  "data": Object {
    "person": Object {
      "name": "Jerry",
      "parents": Object {},
    },
  },
}
`;

exports[`sets node edge to existing node 1`] = `
Object {
  "data": Object {
    "alice": Object {
      "person": Object {
        "name": "Alice",
      },
    },
    "tim": Object {
      "person": Object {
        "name": "Tim",
      },
    },
  },
}
`;

exports[`sets node edge to existing node 2`] = `
Object {
  "data": Object {
    "setPersonPartner": Object {
      "person": Object {
        "name": "Tim",
        "partner": Object {
          "name": "Alice",
          "partner": Object {
            "name": "Tim",
          },
        },
      },
    },
  },
}
`;

exports[`sets node edge to new node 1`] = `
Object {
  "data": Object {
    "tim": Object {
      "person": Object {
        "name": "Tim",
      },
    },
  },
}
`;

exports[`sets node edge to new node 2`] = `
Object {
  "data": Object {
    "setPersonPartner": Object {
      "person": Object {
        "name": "Tim",
        "partner": Object {
          "name": "Wilma",
          "partner": Object {
            "name": "Tim",
          },
        },
      },
    },
  },
}
`;

exports[`sets one way edge to existing node 1`] = `
Object {
  "data": Object {
    "alice": Object {
      "person": Object {
        "name": "Alice",
      },
    },
    "tim": Object {
      "person": Object {
        "name": "Tim",
      },
    },
  },
}
`;

exports[`sets one way edge to existing node 2`] = `
Object {
  "data": Object {
    "setPersonEmergencyContact": Object {
      "person": Object {
        "emergencyContact": Object {
          "name": "Alice",
        },
        "name": "Tim",
      },
    },
  },
}
`;

exports[`setting edge to existing node of incorrect type returns error 1`] = `
Object {
  "data": Object {
    "tim": Object {
      "person": Object {
        "name": "Tim",
      },
    },
    "tom": Object {
      "person": Object {
        "name": "Tom",
      },
    },
  },
}
`;

exports[`setting edge to existing node of incorrect type returns error 2`] = `
Object {
  "data": Object {
    "addPersonEmails": null,
  },
  "errors": Object {
    "0": Object {
      "locations": Object {
        "0": Object {
          "column": 5,
          "line": 2,
        },
      },
      "message": "Cannot create edge \\"emails\\" from \\"Person\\" to \\"Person\\", should be \\"Email\\"",
      "path": Object {
        "0": "addPersonEmails",
      },
    },
  },
}
`;

exports[`setting edge to non-existent node returns error 1`] = `
Object {
  "data": Object {
    "tom": Object {
      "person": Object {
        "name": "Tom",
      },
    },
  },
}
`;

exports[`setting edge to non-existent node returns error 2`] = `
Object {
  "data": Object {
    "setPersonPartner": null,
  },
  "errors": Object {
    "0": Object {
      "locations": Object {
        "0": Object {
          "column": 5,
          "line": 2,
        },
      },
      "message": "There is no \\"Person\\" with id \\"0x00\\"",
      "path": Object {
        "0": "setPersonPartner",
      },
    },
  },
}
`;

exports[`setting edge with no reverse does nothing to existing node 1`] = `
Object {
  "data": Object {
    "alice": Object {
      "person": Object {
        "name": "Alice",
      },
    },
    "tim": Object {
      "person": Object {
        "emergencyContact": Object {
          "name": "Sally",
        },
        "name": "Tim",
      },
    },
  },
}
`;

exports[`setting edge with no reverse does nothing to existing node 2`] = `
Object {
  "data": Object {
    "setPersonEmergencyContact": Object {
      "person": Object {
        "emergencyContact": Object {
          "name": "Alice",
        },
        "name": "Tim",
      },
    },
  },
}
`;

exports[`setting edge with no reverse does nothing to existing node 3`] = `
Object {
  "data": Object {
    "person": Object {
      "emergencyContact": Object {
        "name": "Gerald",
      },
      "name": "Sally",
    },
  },
}
`;

exports[`setting edge with reverse removes reverse edge from existing node 1`] = `
Object {
  "data": Object {
    "alice": Object {
      "person": Object {
        "name": "Alice",
      },
    },
    "tim": Object {
      "person": Object {
        "name": "Tim",
        "partner": Object {
          "name": "Sally",
        },
      },
    },
  },
}
`;

exports[`setting edge with reverse removes reverse edge from existing node 2`] = `
Object {
  "data": Object {
    "setPersonPartner": Object {
      "person": Object {
        "name": "Tim",
        "partner": Object {
          "name": "Alice",
        },
      },
    },
  },
}
`;

exports[`setting edge with reverse removes reverse edge from existing node 3`] = `
Object {
  "data": Object {
    "person": Object {
      "name": "Sally",
      "partner": null,
    },
  },
}
`;

exports[`setting node edge removes existing reverse edge from new value 1`] = `
Object {
  "data": Object {
    "alice": Object {
      "person": Object {
        "name": "Alice",
        "partner": Object {
          "name": "Sue",
        },
      },
    },
    "tim": Object {
      "person": Object {
        "name": "Tim",
      },
    },
  },
}
`;

exports[`setting node edge removes existing reverse edge from new value 2`] = `
Object {
  "data": Object {
    "setPersonPartner": Object {
      "person": Object {
        "name": "Tim",
        "partner": Object {
          "name": "Alice",
        },
      },
    },
  },
}
`;

exports[`setting node edge removes existing reverse edge from new value 3`] = `
Object {
  "data": Object {
    "person": Object {
      "name": "Sue",
      "partner": null,
    },
  },
}
`;

exports[`setting node edge to existing value has no effect 1`] = `
Object {
  "data": Object {
    "tim": Object {
      "person": Object {
        "name": "Tim",
        "partner": Object {
          "name": "Alice",
        },
      },
    },
  },
}
`;

exports[`setting node edge to existing value has no effect 2`] = `
Object {
  "data": Object {
    "setPersonPartner": Object {
      "person": Object {
        "name": "Tim",
        "partner": Object {
          "name": "Alice",
          "partner": Object {
            "name": "Tim",
          },
        },
      },
    },
  },
}
`;

exports[`unsets node edge 1`] = `
Object {
  "data": Object {
    "tim": Object {
      "person": Object {
        "name": "Tim",
        "partner": Object {
          "name": "Joan",
        },
      },
    },
  },
}
`;

exports[`unsets node edge 2`] = `
Object {
  "data": Object {
    "unsetPersonPartner": Object {
      "person": Object {
        "name": "Tim",
        "partner": null,
      },
    },
  },
}
`;

exports[`unsetting node edge removes reverse edge on value 1`] = `
Object {
  "data": Object {
    "tim": Object {
      "person": Object {
        "name": "Tim",
        "partner": Object {
          "name": "Joan",
        },
      },
    },
  },
}
`;

exports[`unsetting node edge removes reverse edge on value 2`] = `
Object {
  "data": Object {
    "unsetPersonPartner": Object {
      "person": Object {
        "name": "Tim",
        "partner": null,
      },
    },
  },
}
`;

exports[`unsetting node edge removes reverse edge on value 3`] = `
Object {
  "data": Object {
    "person": Object {
      "name": "Joan",
      "partner": null,
    },
  },
}
`;
